lab/src/raft/raft.go

```diff
@@ -152,13 +156,36 @@ type RequestVoteArgs struct {
 //
 type RequestVoteReply struct {
 	// Your data here (2A).
+	Term int
+	VoteGranted bool
 }
 
 //
 // example RequestVote RPC handler.
 //
 func (rf *Raft) RequestVote(args *RequestVoteArgs, reply *RequestVoteReply) {
 	// Your code here (2A, 2B).
+	rf.mu.Lock()
+
+	localLastLogIndex := len(rf.log)
+	localLastLogTerm := 0
+	if localLastLogTerm > 0 {
+		localLastLogTerm = rf.log[localLastLogIndex - 1].Term
+	}
+
+	isTermValid := args.Term >= rf.currentTerm
+	isLogUpToDate := 
+		args.LastLogTerm > localLastLogTerm || (args.LastLogTerm == localLastLogTerm && args.LastLogIndex >= localLastLogIndex)
+	isAbleToVote := rf.votedFor == -1 || rf.votedFor == args.CandidateId
+
+	if isTermValid && isLogUpToDate && isAbleToVote {
+		reply.VoteGranted = true
```

Comment:
Do you have to set `rf.votedFor = args.CandidateId` is the vote is granted?

Reply:
Yes, forgot to do this